### ISSUE TITLE
Question / Comment: bbox doesn't exactly match the word

### ISSUE THREAD

**Issue by victor-ab** [pymupdf/PyMuPDF#739]
For some documents (just a few), the bbox doesn't exactly match the word.

```python
doc = fitz.open(pdf_dir)
pages = [ doc[ i ] for i in range( doc.pageCount ) ]
page = pages[0]

blocks = page.getText("dict")['blocks']
text_blocks = [block for block in blocks if block['type']==0]
color='red'
width=2
TARGET_DPI = 300
mat = fitz.Matrix(TARGET_DPI/ 72, TARGET_DPI/ 72)
pix_map = doc.getPagePixmap(0,matrix=mat)
image = Image.open(BytesIO(pix_map.getImageData()))
draw = ImageDraw.Draw(image)

for block in text_blocks:
    for lines in text_blocks:
        for line in lines['lines']:
            for span in line['spans']:
                x0, y0, x1, y1 = span['bbox']
                x0, y0, x1, y1 = x0*TARGET_DPI/ 72, y0*TARGET_DPI/ 72, x1*TARGET_DPI/ 72, y1*TARGET_DPI/ 72
                draw.rectangle((x0, y0, x1, y1), fill=None, outline='green', width=width)
```
![image](https://user-images.githubusercontent.com/26576394/100552167-25f71400-3264-11eb-92da-a972da999ea5.png)

If I add `y0 = y1-span['size']`, then it works:
![image](https://user-images.githubusercontent.com/26576394/100552143-fea04700-3263-11eb-9a94-0be0c5e4ea06.png)
Why does it happens? Is it safer to always use the size?

**Comment by JorjMcKie** [comment]
I guess I don't really understand your code:
Why do you not directly use the span bbox as the coordinates of the rectangle to draw? Instead you enlarge it by 300/72 = 4.17.
If **_the following_** doesn't match, we have a real issue and I would need a reproducer:
`page.drawRect(span["bbox"], width=0.3, color=green)`.

Apart from that argument, I **_have_** seen font examples, which do not correctly return the character bbox (this is the "atom", the larger bboxes for span, line and block are computed by joining these).
The char bbox is computed by MuPDF - which uses font properties for this. If the font has problems, then the result is an incorrect char bbox of course. In the mentioned example, tha char bbox had a height of zero, So my code now checks this and uses the font size in that case ...
To cut a long story short: if my previous instruction produces incorrect rectangles, please send me the example.

**Comment by victor-ab** [comment]
I am sorry, this ratio 300/72 is just for enlarging the dpi and matching it with the bounding box coordinates.
Had the same output with `drawRect`.
[raw-Copy.pdf](https://github.com/pymupdf/PyMuPDF/files/5612910/raw-Copy.pdf)

I edited the pdf with Wondershare PDFelement, but the problem remains:
![image](https://user-images.githubusercontent.com/26576394/100554729-77f46580-3275-11eb-9ef2-f0cc17cf3bca.png)

**Comment by JorjMcKie** [comment]
Indeed font issues. I made an adjustment which ensures that span bboxes can not have a smaller height than the fontsize.
This snippet leads to the following picture:
```python
>>> for block in page.getText("dict",flags=0)["blocks"]:
	for line in block["lines"]:
		for span in line["spans"]:
			_ = page.drawRect(span["bbox"], width=0.3, color=(1,0,0))

```
![grafik](https://user-images.githubusercontent.com/8290722/100579697-08a16480-32bb-11eb-9c2e-ad5f16d18184.png)
An irritation remains:
Your example PDF has two fonts - well-known ones, at least per name. And both show the same problem ...
The following things also require more consideration:
* I applied my fix only to span heights (which percolates up to line an block heights)
* single character bboxes still have a too small height, because I do not want all of them as high as an uppercase character.
* Other extraction types, speficially `getText("words")`, remain to be treated accordingly.

**Comment by victor-ab** [comment]
@JorjMcKie What was the applied change? 
Just adding that flag `0`? If so, it did not worked here.

**Comment by JorjMcKie** [comment]
No it was just "your" fix: making sure the span bbox height is not less than fontsize - however in the C-module behind ``gettext("dict")``.
The flags=0 just make sure that no images are extracted, that ligatures ("fi", "ffi", etc.) are decomposed and that white space is treated as blanks.
I am now developing an algorithm which generalizes the height correction on a by-character level - which is only trivial for horizontal writing mode and zero inclination wrt the x-axis.

**Comment by JorjMcKie** [comment]
Finished a stable version, which resorts to **_"your fix"_** if the character bbox height is smaller than the fontsize.
This version should work correctly if text is in horizontal write mode and rotated by some int multiple of 90 degrees.

This should cover the vast majority of all cases. Also dealing with non-orthogonal text inclinations is far more complicated, allthough certainly possible.
To give you an impression that the problem is the **_font_** (not PyMuPDF / MuPDF) compare the outputs of your page:
* Number 1 with the repaired bbox height. Note how narrow the boxes around text are ...
![grafik](https://user-images.githubusercontent.com/8290722/100744284-f6efb800-33b3-11eb-9404-122153c011a6.png)

* Number 2 after I replaced the font by Helvetica. The boxes are a lot more "generous" now!
![grafik](https://user-images.githubusercontent.com/8290722/100744348-0ff86900-33b4-11eb-98e1-34ff97419036.png)

**Comment by victor-ab** [comment]
Glad to hear you are making progress!

That's also a very interesting aspect about bounding boxes.

Why aren't the bboxes always narrow around the text? Even in a PDF with "good fonts", They always seems to be kinda "off".
Is there an option to control the margin of the bounding box?
![image](https://user-images.githubusercontent.com/26576394/100746469-8058b780-33bf-11eb-8075-79827bee73b5.png)

**Comment by JorjMcKie** [comment]
with the algorithm pinned down in previous post, we get the following "optimized" span bboxes for the well-behaving Helvetica font:
![grafik](https://user-images.githubusercontent.com/8290722/100757423-bea4a580-33c4-11eb-9475-1985eb22cbdb.png)

**Comment by JorjMcKie** [comment]
Watch how "Currency" bbox includes the descending character "y"!
The other bboxes still seem a bit large if they have no descending character. Being awkward, you could resort to "rawdict" and do your own calculation of each character's bbox (fingers crossed that the font is super-wellbehaving and provides individually different such bboxes!) ...

**Comment by victor-ab** [comment]
Hmmm, interesting.

How can I exactly identify which font from the document is being used in a specific span? The fontName extracted from rawdict is kinda generic.
For instance, in another pdf, the fontName in a span is just `Arial`. But when I extract the font list, I get:
```python 
>>> page.getFontList()
[(6, 'ttf', 'Type0', 'IHBWDB+Arial', 'F1', 'Identity-H'),
 (18, 'ttf', 'Type0', 'IHBWDB+Arial', 'F2', 'Identity-H'),
 (26, 'ttf', 'Type0', 'IHBWDB+ArialNarrow', 'F3', 'Identity-H')]
```

But, unfortunately, it is very hard to believe that most of fonts will be "super-wellbehaving".
Here is an example:
```python
for block in text_blocks:
    for lines in text_blocks:
        for line in lines['lines']:
            for span in line['spans']:
                for char in span['chars']:
                    page.drawRect(char['bbox'], width=0.3, color=(0,1,1))
```
![image](https://user-images.githubusercontent.com/26576394/100761263-76d84b00-33d1-11eb-9b1f-036bc0c8f963.png)

**Comment by JorjMcKie** [comment]
Font names starting with 6 upper case letter like `IHBWDB+Arial` are subsets of the same font, the name of which follows the "+".
The two fonts with the same name, but under different xrefs, 6, 18, **must** be the same by PDF spec!
They appear in different xrefs, because that PDF is badly optimized.
The buffers stored for xref 6, resp. 18, **MUST** be the same, or someone made a major error.
To get a clearer picture, you may want to do a ``page.cleanContents()`` before making a fontlist or extract text.

I once did include that 6-letter prefix in getText() but dropped it again at some point, b/o no value added information.
And you should receive "ArialNarrow" if that font is indeed used on the page - which need not be the case just because that font appears in the list!
So, what you could do is searching the fontlist for a name ending with the getText() fontname - sometimes a more sophisticated search is required, because you are actually looking for the name of the basefont ...
Then extract the fontbuffer via the xref and make a temporary font to find out those ascender/descender properties.

Well, and if all fails, and the font contains no character-specific bbox, you will have to live with the font level one. The info simply does not exist then.

>  In this case, even the y0 is off.

What do you mean, the picture looks ok to me.

**Comment by victor-ab** [comment]
Got it, thanks a lot for the comprehensive explanation!
I guess it will be better to just live with that. If two bboxes overlaps, I'll just change move overlapping boundary to the middle of that overlap.


> In this case, even the y0 is off.

I meant that the `y0` is higher than the upper-case character.

**Comment by JorjMcKie** [comment]
This is one of those days with sh…y internet connection. So trying to e-mail the response.

I think I should return those two properties, ascender / descender, as part of the “dict” / rawdict” dictionaries. MuPDF is handing the font in – I just do not use those two values yet.

**Comment by JorjMcKie** [comment]
The next PyMuPDF version (1.18.5) will have this as discussed:
```python
>>> pprint(page.getText("dict",flags=0)["blocks"])
[{'bbox': (100.0, 88.17500305175781, 117.73199462890625, 103.28900146484375),
  'lines': [{'bbox': (100.0,
                      88.17500305175781,
                      117.73199462890625,
                      103.28900146484375),
             'dir': (1.0, 0.0),
             'spans': [{'ascender': 1.0750000476837158,  # <=====
                        'bbox': (100.0,
                                 88.17500305175781,
                                 117.73199462890625,
                                 103.28900146484375),
                        'color': 0,
                        'descender': -0.29899999499320984,  # <=====
                        'flags': 0,
                        'font': 'Helvetica',
                        'origin': (100.0, 100.0),
                        'size': 11.0,
                        'text': 'text'}],
             'wmode': 0}],
  'number': 0,
  'type': 0}]
>>> 
```

**Comment by JorjMcKie** [comment]
If you want I can provide a pre-version to test ...
Would need your ``print(fitz.__doc__)`` output.

**Comment by victor-ab** [comment]
Yes, I'd like to try it.

```python
>>> print(fitz.__doc__)
PyMuPDF 1.18.4: Python bindings for the MuPDF 1.18.0 library.
Version date: 2020-11-19 08:56:23.
Built for Python 3.7 on win32 (64-bit).
```

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.18.5-cp37-cp37m-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/5624648/PyMuPDF-1.18.5-cp37-cp37m-win_amd64.zip)
rename extension from zip to whl an then use pip install -U with it

**Comment by victor-ab** [comment]
```python

def recalculate_bbox(span):
    a = span["ascender"] / 1000 * span["size"]
    d = span["descender"] / 1000 * span["size"]
    o = fitz.Point(span["origin"])
    r = fitz.Rect(span["bbox"])
    r.y1 = o.y - span["size"] * d / (a - d)
    r.y0 = r.y1 - span["size"]
    return r

for block in text_blocks:
    for lines in text_blocks:
        for line in lines['lines']:
            for span in line['spans']:
                page.drawRect(recalculate_bbox(span), width=0.3, color=(0,1,1))
```
Produced:
![image](https://user-images.githubusercontent.com/26576394/100783628-caa45d80-33ec-11eb-9f9b-cd6a9e2fbe83.png)
This is way better now. Already perfect for my application.

**Comment by JorjMcKie** [comment]
One additional comment: the ascender / descender value in the dict are already floats divided by 1000.
So you do not need to do that again in your function.
Relevant for calculating the correcting is only the quotient `d/(a-d)`. So the following would do just the same job:
```python
def recalculate_bbox(span):
    a = span["ascender"]
    d = span["descender"]
    o = fitz.Point(span["origin"])
    r = fitz.Rect(span["bbox"])
    r.y1 = o.y - span["size"] * d / (a - d)
    r.y0 = r.y1 - span["size"]
    return r
```

This was a very satisfying discussion! Thanks for bringing this up!

**Comment by JorjMcKie** [comment]
Solved by v1.18.5 being uploaded now.
